aten/src/ATen/native/vulkan/VulkanAten.cpp

```diff
@@ -0,0 +1,297 @@
+
+#include <ATen/ATen.h>
+#include <ATen/Config.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/OpaqueTensorImpl.h>
+#include <ATen/native/UpSample.h>
+#include <ATen/native/utils/ParamUtils.h>
+#include <ATen/native/vulkan/VulkanDebugUtils.h>
+
+#ifdef USE_VULKANGL
+#include <ATen/native/vulkan/VulkanGL.h>
+#endif
+
+#ifdef USE_VULKAN
+#include <ATen/native/vulkan/VulkanVulkan.h>
+#endif
+
+namespace at {
+namespace native {
+
+template <typename T>
+struct CAFFE2_API IntrusivePtrTargetWrapper : c10::intrusive_ptr_target {
+ private:
+  T target_;
+
+ public:
+  IntrusivePtrTargetWrapper() = delete;
+  IntrusivePtrTargetWrapper(const T& target) : target_(target) {}
+  IntrusivePtrTargetWrapper(T&& target) : target_(std::move(target)) {}
+
+  T& get_target() {
+    return target_;
+  }
+};
+
+#ifdef USE_VULKAN
+using VTensor = at::native::vulkan::details::vulkan::VulkanVulkanTensor;
+#endif
+#ifdef USE_VULKANGL
+using VTensor = at::native::vulkan::details::gl::VulkanGLTensor;
+#endif
+
+using VTensorWrapper = IntrusivePtrTargetWrapper<VTensor>;
+using VTensorWrapperPtr = c10::intrusive_ptr<VTensorWrapper>;
+using VulkanTensorImpl = OpaqueTensorImpl<VTensorWrapperPtr>;
+using VulkanTensor = at::Tensor;
+
+std::ostream& operator<<(std::ostream& s, const VulkanTensor& phvTensor) {
```

Comment:
> What is "phv"?
Just artifact from my sandbox, it was my acronym for "pytorch vulkan". I will clean it.

> Since VulkanTensor is just an alias for at::Tensor, doesn't this collide with the existing overload for at::Tensor

Yes, we can put at::Tensor, but the same as in mkldnn approach where MKLDNNTensor was introduced as at::Tensor alias.

But here I think I expected that overloading of printing of tensor will be dispatched here as perfect alias match, but I can be wrong.